topics/07_web_ui_testing/main_web_automation_with_playwright.md

```diff
@@ -1,3 +1,383 @@
+5. Easy to use within CI/CD (e.g. Jenkins)
+
+## Remark
+Playwright is one of numerous test frameworks that use/support Python
+Some other popular Python Test Frameworks:
+1. Robot Test Framework: powerful frameworks that can be used without deep knowledge of Python. It allows creating tests that are convenient for reading by non-technical stakeholders. Website: robotframework.org
+2. Pytest: versatile testing framework supporting unit, functional, and integration testing. Website: pytest.org
+3. Behave: A behavior-driven development (BDD) framework. Website: https://behave.readthedocs.io
+
+## References:
+
+1. [Playwright - Python official page](https://playwright.dev/python/)
+2. [Installation](https://playwright.dev/python/docs/intro)
+3. [Assertion](https://playwright.dev/python/docs/test-assertions)
+
+
+## Playwright Installation:
+
+From Python point of view Playwright is just a package
+
+Installation: 
+```python
+pip install pytest-playwright
+```
+
+## Playwright Launching:
+
+The simplest launching command: 
+```python
+pytest filename_with_playwright_code.py
+```
+
+# Framework structure
+1 Service classes
+2 Page object
+3 Test suits
+
+## Test suit top level components:
+
+1. Import libraries
+2. Import page Page Object(s)
+3. Test
+- Action
+- Assertion
+
+## Page object top level components:
+1. Import libraries
+2. PO constructor (including locators)
+3. Methods of PO
+- Interacting with page
+- Getting information
+- Assertion methods 
+
+### PageObject
+
+PageObject is the kernel of Playwright UI test
+
+1. Import libraries
+Page Object Structure example: 
+```python
+from playwright.sync_api import Page
+
+2. PO constructor (including locators)
+class PoClassNamePage:
+    # Initialize page and selectors
+    def __init__(self, page: Page):
+        self.page = page
+        # list of selectors
+        self.fullNameField = page.locator("#userName")
+        self.emailField = page.locator("#userEmail")
+        self.currentAddressField = page.locator("#currentAddress")
+        self.permanentAddressField = page.locator("#permanentAddress")
+        self.submitButton = page.locator("#submit")
+        self.pageLoadingIndicator = page.locator(".text-center")
+
+3. Methods of PO
+    # method to iteract with page
+    def navigate(self, url: str):
+        self.page.goto(url)
+        self.currentAddressField.click()
+
+    # method to get information
+    def navigate(self, url: str):
+        self.permanentAddressField.text_content()
+
+    # Assertion methods
+    def check_loading_page(self):
+        return self.fullNameField.text_content() == "John Smith"
+```
+> [!TIP]
+> locator can be organized like 'chain':
+
+```python
+self.access_block = page.locator("//*div[@id='add-visibility-form']")
+self.access_block_select_type = self.access_block.locator("/div[@id='id_requied_action']") 
+instead of "//*div[@id='add-visibility-form']/div[@id='id_requied_action']"
+```
+
+### Assertions
```

Comment:
Try to format the words (bold, italic etc) and code (code block using the triple backticks) to better highlighting things.

Reply:
Partly updated